internal/conf/load.go

```diff
@@ -0,0 +1,155 @@
+// INTECH Process Automation Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package conf
+
+import (
+	"fmt"
+	"github.com/mitchellh/mapstructure"
+	"gopkg.in/yaml.v3"
+	"io/ioutil"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+)
+
+const Separator = "__"
+const Prefix = "KUIPER"
+
+func LoadConfig(c interface{}) error {
+	return LoadConfigByName(ConfFileName, c)
+}
+
+func LoadConfigByName(name string, c interface{}) error {
+	dir, err := GetConfLoc()
+	if err != nil {
+		return err
+	}
+	p := path.Join(dir, name)
+	return LoadConfigFromPath(p, c)
+}
+
+func LoadConfigFromPath(path string, c interface{}) error {
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	configMap := make(map[string]interface{})
+	err = yaml.Unmarshal(b, &configMap)
+	if err != nil {
+		return err
+	}
+	var configs map[string]interface{}
+	if _, success := c.(*map[string]interface{}); success {
+		configs = configMap
```

Comment:
I suppose here is the case of source config reading. Even read into map, we still need to deal with case. For example, edgex.yaml has a key "messageType" in yaml and after reading the map key is also camel case "messageType"; However, the env var will be parsed to "messagetype". How do we deal with this right now?

Reply:
maybe we can get the key case from its json file, for example edgex.json has a edgex.json file with properties section that listed all the available properties.

Reply:
@ngjaying I disabled normalization for map type config due to one of failing tests. Maybe we should fix the test instead of creating an edge case for the map type ?

Reply:
There is actually a big difference here. For struct, we can implicitly define the case mapping by the yaml tag. But for map, there is no tag. How do you know what an env var MESSAGETYPE should be converted to?

Reply:
I think it would good to make the same behavior - make everything case insensitive. Somehow it does not convince me that for yaml we should be case insensitive while for map case sensitive.

Reply:
yaml is case sensitive now as we read it from `yaml.Unmarshal(b, &output)`. We did not want to change this behavior. When the output is a struct, the case sensitive yaml key are map to the struct field with the help of yaml tag of the struct. Othwise, if the outpus is a map, the case sensitive yaml key directly become the map key. For example, yaml key `messageType` will become `MessageType` in a struct and become `messageType` in a map. 

The problem is that the env is case insensitive and you normalize key to handle it. If the output is a struct, we can normalize the key and changes back to the case sensitive struct key with the help of mapstructure decode. But for map, if we normalize the key, we cannot revert back. We seems to miss the key case information. And that's my concern.

Reply:
Moreover, the struct always define all available keys(fields), but map we only read from yaml. If the yaml does not define all keys, we have no way to know if a key from ENV is valid or not. I think the root cause is the missing information of property meta data for source configure. We can actually read that meta data from an optional json file like edgex.json.

Reply:
> Moreover, the struct always define all available keys(fields), but map we only read from yaml. If the yaml does not define all keys, we have no way to know if a key from ENV is valid or not. I think the root cause is the missing information of property meta data for source configure. We can actually read that meta data from an optional json file like edgex.json.

You mean add such functionality - json to map[string]interface{} ?

Reply:
No. There is an optional json file for each plugin. For example, etc/sources/edgex.json defines the meta data of edgex source.  Inside it, there is a section properties which define all the available properties with the correct case. When parsing yaml to map we can use the properties information to validate the keys. When convert ENV to conf, we can also use the properties information to map the key from upper case to the correct case.

Reply:
It is an array - how does this refer to config ?

Reply:
Let me take edgex as an example. From etc/sources/edgex.json, we can get a list of correct case properties(keys) like ["protocol", "messageType"...]. In internal/topo/node/node.go line 182, you parse yaml to a map. At there, we can loop the map key and validate against the key array of json. Then in this file, when handling env, we can still normalize the key and when decoding back to map, use the key array of json to restore the case of the keys.

Reply:
ok I think I finally understood what you are writing about, where I am supposed to look for these files. One file at particular ocassion will suffice or particular yaml requires particular json ?

Reply:
Great! Each yaml should have a json file alongside for all built-in sources. It is optional for extended plugin, so if json is not found, we can just skip this check.

Reply:
Awesome, are these json locations the same as yaml counterparts or they have specific location ?

Reply:
yes, they are at the same location.

Reply:
hi, I made the changes but I have problems [here](https://github.com/lf-edge/ekuiper/runs/3549255802?check_suite_focus=true), can you help me ?

Reply:
Hi Robert, that should be a network problem of the tdengine client download. We are aware of it and just check with the tdengine team. Let's just leave it there.

Reply:
What about changes, is this you were writing about ?

Reply:
Yes, thanks